boot/freeldr/freeldr/lib/fs/fs.c

```diff
@@ -43,162 +43,221 @@ typedef struct tagDEVICE
 {
     LIST_ENTRY ListEntry;
     const DEVVTBL* FuncTable;
-    CHAR* Prefix;
+    PSTR DeviceName;
     ULONG DeviceId;
     ULONG ReferenceCount;
 } DEVICE;
 
 static FILEDATA FileData[MAX_FDS];
 static LIST_ENTRY DeviceListHead;
 
+#define IS_VALID_FILEID(FileId) \
+    ((FileId) < _countof(FileData) && FileData[(FileId)].FuncTable)
+
+typedef const DEVVTBL* (*PFS_MOUNT)(ULONG DeviceId);
+
+PFS_MOUNT FileSystems[] =
+{
+#ifndef _M_ARM
+    IsoMount,
+#endif
+    FatMount,
+    BtrFsMount,
+#ifndef _M_ARM
+    NtfsMount,
+    Ext2Mount,
+#endif
+#if defined(_M_IX86) || defined(_M_AMD64)
+#ifndef UEFIBOOT
+    PxeMount,
+#endif
+#endif
+};
+
+
 /* ARC FUNCTIONS **************************************************************/
 
+/**
+ * @brief
+ * Replace "()" by "(0)" in the given ARC device name, if necessary.
+ *
+ * @param[in]   DeviceName
+ * The original ARC device name string to normalize.
+ *
+ * @param[in,out]   Length
+ * Points to a SIZE_T variable that:
+ * - on input, specifies the length of the ARC device name string pointed
+ *   by DeviceName;
+ * - on output, receives the length of the normalized ARC device name,
+ *   returned by value by this function.
+ *
+ * @return
+ * - On success, returns a string pointer to the normalized ARC device name.
+ *   If the device name string had to be normalized, this returned string is
+ *   allocated from the temporary heap, with TAG_DEVICE_NAME tag.
+ *   Otherwise, the function returns the original device name string pointer.
+ * - On failure, the function returns NULL.
+ **/
+PCHAR
+NormalizeArcDeviceName(
+    _In_ PCCH DeviceName,
+    _Inout_ PSIZE_T Length)
+{
+    ULONG Count;
+    SIZE_T NameLength;
+    PCCH p, End;
+
+    NameLength = *Length;
+    End = DeviceName + NameLength;
+
+    /* Count the number of "()", which needs to be replaced by "(0)" */
+    Count = 0;
+    for (p = DeviceName; p < End; ++p)
+    {
+        if ((p + 1) < End && *p == '(' && *(p + 1) == ')')
+            ++Count; //, ++p;
+    }
+
+    /* Duplicate the device name, replacing "()" by "(0)" if necessary */
+    *Length = NameLength + Count;
+    if (Count != 0)
```

Comment:
@hpoussin 's suggestion:
Doing `if (Count == 0) return (PCHAR)DeviceName;` in order to reduce the level of indentation.